packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +50,430 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
+    );
+  } else if (type === "object") {
+    assert(isUserDefined(objectType), errMessage(name, "A user-defined type must be specified through 'objectType'."));
+  } else if (type === "linkingObjects") {
+    assert(isUserDefined(objectType), errMessage(name, "A user-defined type must be specified through 'objectType'."));
+    assert(!!property, errMessage(name, "The linking object's property name must be specified through 'property'."));
   } else {
-    return { type: "object", objectType: type, optional: true };
+    // 'type' is a user-defined type
+    error(
+      name,
+      `If you meant to define a relationship, use { type: 'object', objectType: '${type}' } or { type: 'linkingObjects', objectType: '${type}', property: 'The ${type} property' }`,
+    );
+  }
+
+  if (optionalIsImplicitlyTrue(type, objectType)) {
+    const displayed =
+      type === "mixed" || objectType === "mixed"
+        ? "'mixed' types"
+        : "user-defined types as standalone objects and in dictionaries";
+    assert(optional !== false, errMessage(name, `'optional' is always 'true' for ${displayed} and cannot be set to 'false'.`));
+    optional = true;
+  } else if (optionalIsImplicitlyFalse(type, objectType)) {
+    const displayed = type === "linkingObjects" ? "linking objects" : "user-defined types in lists and sets";
+    assert(optional !== true, errMessage(name, `'optional' is always 'false' for ${displayed} and cannot be set to 'true'.`));
+    optional = false;
+  }
+
+  const normalizedSchema: CanonicalObjectSchemaProperty = {
+    name: name.propertyName,
+    type: type as PropertyTypeName,
+    optional: !!optional,
+    indexed: !!schema.indexed,
+    mapTo: schema.mapTo || name.propertyName,
+  };
+  // Add optional properties only if defined (tests expect no 'undefined' properties)
+  objectType !== undefined && (normalizedSchema.objectType = objectType);
+  property !== undefined && (normalizedSchema.property = property);
+  defaultValue !== undefined && (normalizedSchema.default = defaultValue);
+
+  return normalizedSchema;
+}
+
+/**
+ * Sanitize the top-level fields of a user-provided object schema by
+ * validating the data types.
+ */
+export function sanitizeObjectSchema(schema: unknown): asserts schema is ObjectSchema {
+  assert.object(schema, "the object schema", false);
+  assert.string(schema.name, "the object schema name");
+  assert.object(schema.properties, `${schema.name}.properties`, false);
+  if (schema.primaryKey !== undefined) {
+    assert.string(schema.primaryKey, `${schema.name}.primaryKey`);
+  }
+  if (schema.embedded !== undefined) {
+    assert.boolean(schema.embedded, `${schema.name}.embedded`);
+  }
+  if (schema.asymmetric !== undefined) {
+    assert.boolean(schema.asymmetric, `${schema.name}.asymmetric`);
+  }
+  const allowedKeys = new Set<keyof ObjectSchema>(["name", "primaryKey", "embedded", "asymmetric", "properties"]);
```

Comment:
Not that this really matters but consider moving this declaration out of the function. At least in theory, this creates a new set for each call, but we don't need that. A good optimizer should avoid that, but I have no idea if JS engines do that optimization.

On the other hand, this code is the opposite of perf sensitive since it runs roughly once per app, and this is a cheap operation anyway. So if you decide to leave it like this after considering it, that is fine with me. This is just one of those things my eyes have been trained to notice and I can't help but point it out.

Reply:
Even though it's ~1/app, I agree that it should be moved outside 👍